File: src/state.ts
import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { deserialize, serialize, field, vec } from 'borsh';

export const MAX_ROUTE_HOPS = 5;

export class RouteState {
  @field({ type: 'u8' })
  discriminator: number = 0;

  @field({ type: 'publicKey' })
  inputMint: PublicKey = PublicKey.default;

  @field({ type: 'publicKey' })
  outputMint: PublicKey = PublicKey.default;

  @field({ type: 'u64' })
  amountIn: BN = new BN(0);

  @field({ type: 'u64' })
  minAmountOut: BN = new BN(0);

  @field({ type: 'u8' })
  hops: number = 0;

  @field({ type: vec('publicKey') })
  route: PublicKey[] = [];

  @field({ type: 'u8' })
  status: number = 0;

  @field({ type: 'publicKey' })
  authority: PublicKey = PublicKey.default;

  constructor(fields?: {
    discriminator?: number;
    inputMint?: PublicKey;
    outputMint?: PublicKey;
    amountIn?: BN;
    minAmountOut?: BN;
    hops?: number;
    route?: PublicKey[];
    status?: number;
    authority?: PublicKey;
  }) {
    if (fields) {
      Object.assign(this, fields);
    }
  }

  static fromBuffer(buffer: Buffer): RouteState {
    return deserialize(RouteState, buffer);
  }

  toBuffer(): Buffer {
    return Buffer.from(serialize(this));
  }
}

export class PoolInfoState {
  @field({ type: 'publicKey' })
  address: PublicKey = PublicKey.default;

  @field({ type: 'publicKey' })
  tokenA: PublicKey = PublicKey.default;

  @field({ type: 'publicKey' })
  tokenB: PublicKey = PublicKey.default;

  @field({ type: 'u16' })
  feeBps: number = 0;

  @field({ type: 'u64' })
  reserveA: BN = new BN(0);

  @field({ type: 'u64' })
  reserveB: BN = new BN(0);

  constructor(fields?: {
    address?: PublicKey;
    tokenA?: PublicKey;
    tokenB?: PublicKey;
    feeBps?: number;
    reserveA?: BN;
    reserveB?: BN;
  }) {
    if (fields) {
      Object.assign(this, fields);
    }
  }

  getOutputAmount(inputMint: PublicKey, amountIn: BN): BN | null {
    const isTokenA = inputMint.equals(this.tokenA);
    const isTokenB = inputMint.equals(this.tokenB);

    if (!isTokenA && !isTokenB) {
      return null;
    }

    const reserveIn = isTokenA ? this.reserveA : this.reserveB;
    const reserveOut = isTokenA ? this.reserveB : this.reserveA;

    // Calculate output: (amountIn * (10000 - feeBps) * reserveOut) / (reserveIn * 10000 + amountIn * (10000 - feeBps))
    const feeFactor = new BN(10000 - this.feeBps);
    const amountInWithFee = amountIn.mul(feeFactor);
    const numerator = amountInWithFee.mul(reserveOut);
    const denominator = reserveIn.mul(new BN(10000)).add(amountInWithFee);

    return numerator.div(denominator);
  }

  getOtherToken(token: PublicKey): PublicKey | null {
    if (token.equals(this.tokenA)) {
      return this.tokenB;
    } else if (token.equals(this.tokenB)) {
      return this.tokenA;
    }
    return null;
  }

  static fromBuffer(buffer: Buffer): PoolInfoState {
    return deserialize(PoolInfoState, buffer);
  }
}

File: program/src/state.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::pubkey::Pubkey;

pub const MAX_ROUTE_HOPS: usize = 5;
pub const ROUTE_STATE_SIZE: usize = 8 + 32 + 32 + 8 + 8 + 1 + (MAX_ROUTE_HOPS * 32) + 1 + 32;

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct RouteState {
    /// Discriminator for account type
    pub discriminator: u8,
    
    /// Input token mint
    pub input_mint: Pubkey,
    
    /// Output token mint
    pub output_mint: Pubkey,
    
    /// Input amount
    pub amount_in: u64,
    
    /// Minimum output amount
    pub min_amount_out: u64,
    
    /// Number of hops in the route
    pub hops: u8,
    
    /// Route path (pool pubkeys)
    pub route: Vec<Pubkey>,
    
    /// Route status: 0 = uninitialized, 1 = initialized, 2 = route_found, 3 = executed
    pub status: u8,
    
    /// Authority
    pub authority: Pubkey,
}

impl RouteState {
    pub const LEN: usize = ROUTE_STATE_SIZE;
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
pub struct PoolInfo {
    /// Pool account address
    pub address: Pubkey,
    
    /// Token A mint
    pub token_a: Pubkey,
    
    /// Token B mint
    pub token_b: Pubkey,
    
    /// Fee in basis points (1 bps = 0.01%)
    pub fee_bps: u16,
    
    /// Token A reserve
    pub reserve_a: u64,
    
    /// Token B reserve
    pub reserve_b: u64,
}

impl PoolInfo {
    pub fn get_output_amount(&self, input_mint: &Pubkey, amount_in: u64) -> Option<u64> {
        let (reserve_in, reserve_out) = if input_mint == &self.token_a {
            (self.reserve_a, self.reserve_b)
        } else if input_mint == &self.token_b {
            (self.reserve_b, self.reserve_a)
        } else {
            return None;
        };

        // Calculate output with fee: (amount_in * (10000 - fee_bps) * reserve_out) / (reserve_in * 10000 + amount_in * (10000 - fee_bps))
        let amount_in_with_fee = (amount_in as u128)
            .checked_mul((10000u128).checked_sub(self.fee_bps as u128)?)?;
        
        let numerator = amount_in_with_fee.checked_mul(reserve_out as u128)?;
        let denominator = (reserve_in as u128)
            .checked_mul(10000u128)?
            .checked_add(amount_in_with_fee)?;
        
        Some((numerator / denominator) as u64)
    }

    pub fn get_other_token(&self, token: &Pubkey) -> Option<Pubkey> {
        if token == &self.token_a {
            Some(self.token_b)
        } else if token == &self.token_b {
            Some(self.token_a)
        } else {
            None
        }
    }
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct PoolRegistry {
    /// Discriminator
    pub discriminator: u8,
    
    /// Authority
    pub authority: Pubkey,
    
    /// Registered pools
    pub pools: Vec<PoolInfo>,
}

